File: day10/haskell/part2.hs
import System.Environment (getArgs)
import Data.List (foldl', sortBy)
import Data.Ord (comparing)

splitOn :: Char -> String -> [String]
splitOn _ [] = []
splitOn delim s = case break (== delim) s of
  (a, []) -> [a]
  (a, _:rest) -> a : splitOn delim rest

extractParens :: String -> [String]
extractParens [] = []
extractParens s = case dropWhile (/= '(') s of
  [] -> []
  (_:rest) -> takeWhile (/= ')') rest : extractParens (dropWhile (/= ')') rest)

extractCurly :: String -> String
extractCurly s = takeWhile (/= '}') $ drop 1 $ dropWhile (/= '{') s

-- Simple brute force for small instances, Gaussian elimination for larger
solve :: [[Int]] -> [Int] -> Int
solve buttons targets
  | m <= 6 = bruteForce buttons targets
  | otherwise = gaussianSolve buttons targets
  where m = length buttons

bruteForce :: [[Int]] -> [Int] -> Int
bruteForce buttons targets = minimum [s | s <- allSums, s >= 0]
  where
    n = length targets
    m = length buttons
    maxT = maximum targets

    allSums = map tryCombo allCombos

    allCombos = sequence [[0..maxT] | _ <- [1..m]]

    tryCombo presses =
      let counters = [sum [presses !! j | j <- [0..m-1], i `elem` (buttons !! j)] | i <- [0..n-1]]
      in if counters == targets then sum presses else maxT * m + 1

gaussianSolve :: [[Int]] -> [Int] -> Int
gaussianSolve buttons targets = best
  where
    n = length targets
    m = length buttons
    maxT = maximum targets

    -- Build augmented matrix
    matrix0 = [[if j < m then if i `elem` (buttons !! j) then 1 else 0 else targets !! i
               | j <- [0..m]] | i <- [0..n-1]]

    -- Gaussian elimination
    (matrix, pivots) = gaussElim matrix0 0 0 []

    gaussElim mat row col piv
      | col >= m || row >= n = (mat, piv)
      | otherwise = case findPivotRow mat row col of
          Nothing -> gaussElim mat row (col+1) piv
          Just pr ->
            let mat' = swapRows mat row pr
                mat'' = eliminate mat' row col
            in gaussElim mat'' (row+1) (col+1) ((col, row):piv)

    findPivotRow mat r c =
      case filter (\i -> (mat !! i) !! c /= 0) [r..n-1] of
        [] -> Nothing
        (i:_) -> Just i

    swapRows mat r1 r2
      | r1 == r2 = mat
      | otherwise = [if i == r1 then mat !! r2 else if i == r2 then mat !! r1 else mat !! i | i <- [0..n-1]]

    eliminate mat r c = [if i == r then row else elimRow row (mat !! r) c | (i, row) <- zip [0..] mat]

    elimRow row pivRow c
      | row !! c == 0 = row
      | otherwise = [row !! j * pv - sv * pivRow !! j | j <- [0..m]]
      where pv = pivRow !! c; sv = row !! c

    -- Free variables
    pivotCols = map fst pivots
    freeVars = [j | j <- [0..m-1], j `notElem` pivotCols]
    numFree = length freeVars

    -- Search over free variables
    best = if numFree > 4
           then searchLimited
           else searchAll

    searchAll = minimum $ map evalFreeVals allFreeVals
      where
        allFreeVals = sequence [[0..maxT] | _ <- [1..numFree]]

    searchLimited = minimum $ map evalFreeVals limitedFreeVals
      where
        limitedFreeVals = sequence [[0..min 20 maxT] | _ <- [1..numFree]]

    evalFreeVals freeVals =
      case backSub freeVals of
        Nothing -> maxT * m + 1
        Just sol -> sum sol

    backSub freeVals = go (sortBy (comparing (negate . snd)) pivots) sol0
      where
        sol0 = [if j `elem` freeVars then freeVals !! indexOf j freeVars else 0 | j <- [0..m-1]]
        indexOf x xs = length $ takeWhile (/= x) xs

        go [] sol = Just sol
        go ((col, row):rest) sol =
          let r = matrix !! row
              pv = r !! col
              rhs = (r !! m) - sum [r !! j * sol !! j | j <- [0..m-1], j /= col]
          in if pv == 0 || rhs `mod` pv /= 0 then Nothing
             else let v = rhs `div` pv
                  in if v < 0 then Nothing
                     else go rest (take col sol ++ [v] ++ drop (col+1) sol)

parseLine :: String -> Int
parseLine line = solve buttons targets
  where
    buttonStrs = extractParens line
    parseButton str = map (\s -> read s :: Int) $ splitOn ',' str
    buttons = map parseButton buttonStrs
    targetStr = extractCurly line
    targets = map (\s -> read s :: Int) $ splitOn ',' targetStr

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "input.txt" else head args
  content <- readFile filename
  let total = sum $ map parseLine $ lines content
  print total

File: day10/go/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func solve(buttons [][]int, targets []int) int {
	n := len(targets)  // counters
	m := len(buttons)  // buttons

	matrix := make([][]int, n)
	for i := 0; i < n; i++ {
		matrix[i] = make([]int, m+1)
		matrix[i][m] = targets[i]
	}
	for j, button := range buttons {
		for _, idx := range button {
			if idx < n {
				matrix[idx][j] = 1
			}
		}
	}

	// Gaussian elimination (integer arithmetic)
	pivotCols := make([]int, 0)
	pivotRows := make([]int, 0)
	row := 0

	for col := 0; col < m && row < n; col++ {
		// Find pivot
		pivotRow := -1
		for i := row; i < n; i++ {
			if matrix[i][col] != 0 {
				pivotRow = i
				break
			}
		}
		if pivotRow == -1 {
			continue // no pivot in this column
		}

		// Swap rows
		matrix[row], matrix[pivotRow] = matrix[pivotRow], matrix[row]

		// Eliminate
		for i := 0; i < n; i++ {
			if i != row && matrix[i][col] != 0 {
				// Subtract matrix[i][col]/matrix[row][col] * row from row i
				// To stay in integers, multiply row i by matrix[row][col] first
				mult := matrix[row][col]
				sub := matrix[i][col]
				for j := 0; j <= m; j++ {
					matrix[i][j] = matrix[i][j]*mult - sub*matrix[row][j]
				}
			}
		}

		pivotCols = append(pivotCols, col)
		pivotRows = append(pivotRows, row)
		row++
	}

	// Identify free variables (non-pivot columns)
	isPivot := make([]bool, m)
	for _, col := range pivotCols {
		isPivot[col] = true
	}
	freeVars := make([]int, 0)
	for j := 0; j < m; j++ {
		if !isPivot[j] {
			freeVars = append(freeVars, j)
		}
	}

	numFree := len(freeVars)
	maxTarget := 0
	for _, t := range targets {
		maxTarget = max(maxTarget, t)
	}

	best := -1

	// Search over free variable values
	var searchFree func(idx int, freeVals []int)
	searchFree = func(idx int, freeVals []int) {
		if idx == numFree {
			// Compute basic variables from free variables
			solution := make([]int, m)
			for i, fv := range freeVars {
				solution[fv] = freeVals[i]
			}

			// Back-substitute to find basic variables
			valid := true
			for i := len(pivotCols) - 1; i >= 0 && valid; i-- {
				r := pivotRows[i]
				c := pivotCols[i]

				// matrix[r][c] * solution[c] + sum of other terms = matrix[r][m]
				rhs := matrix[r][m]
				for j := 0; j < m; j++ {
					if j != c {
						rhs -= matrix[r][j] * solution[j]
					}
				}
				if rhs%matrix[r][c] != 0 {
					valid = false
				} else {
					solution[c] = rhs / matrix[r][c]
					if solution[c] < 0 {
						valid = false
					}
				}
			}

			if valid {
				sum := 0
				for _, v := range solution {
					sum += v
				}
				if best < 0 || sum < best {
					best = sum
				}
			}
			return
		}

		// Prune based on current best
		for v := 0; v <= maxTarget; v++ {
			if best >= 0 {
				// Current sum of free vars
				currentSum := v
				for i := 0; i < idx; i++ {
					currentSum += freeVals[i]
				}
				if currentSum >= best {
					break
				}
			}
			freeVals[idx] = v
			searchFree(idx+1, freeVals)
		}
	}

	if numFree == 0 {
		// No free variables - unique solution
		searchFree(0, []int{})
	} else if numFree <= 5 {
		// Few free variables - search
		searchFree(0, make([]int, numFree))
	} else {
		// Too many free variables - try a different approach
		// Just try small values
		searchFree(0, make([]int, numFree))
	}

	return best
}

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}
	file, _ := os.Open(filename)
	defer file.Close()

	totalPresses := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		var buttons [][]int
		inParen := false
		var current string
		for _, c := range line {
			if c == '(' {
				inParen = true
				current = ""
			} else if c == ')' {
				inParen = false
				var indices []int
				for _, s := range strings.Split(current, ",") {
					idx, _ := strconv.Atoi(s)
					indices = append(indices, idx)
				}
				buttons = append(buttons, indices)
			} else if inParen {
				current += string(c)
			}
		}

		var targets []int
		inCurly := false
		current = ""
		for _, c := range line {
			if c == '{' {
				inCurly = true
				current = ""
			} else if c == '}' {
				inCurly = false
				for _, s := range strings.Split(current, ",") {
					val, _ := strconv.Atoi(s)
					targets = append(targets, val)
				}
			} else if inCurly {
				current += string(c)
			}
		}

		result := solve(buttons, targets)
		if result >= 0 {
			totalPresses += result
		}
	}

	fmt.Println(totalPresses)
}
